src/main/java/org/opensearch/neuralsearch/stats/events/EventStatsManager.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.stats.events;
+
+import lombok.Getter;
+import org.opensearch.neuralsearch.settings.NeuralSearchSettingsAccessor;
+import org.opensearch.neuralsearch.stats.common.StatSnapshot;
+
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Manager class for event counter stats, used to track monotonically increasing counts plus some possible metadata
+ * The statistics collection can be enabled or disabled via cluster settings.
+ */
+public class EventStatsManager {
+    @Getter
+    private Map<EventStatName, EventStat> stats;
+
+    private static EventStatsManager INSTANCE;
+    private NeuralSearchSettingsAccessor settingsAccessor;
+
+    /**
+     * Returns the singleton instance of EventStatsManager.
+     * Creates a new instance with default settings if one doesn't exist.
+     *
+     * @return The singleton instance of EventStatsManager
+     */
+    public static EventStatsManager instance() {
+        if (INSTANCE == null) {
+            INSTANCE = new EventStatsManager(NeuralSearchSettingsAccessor.instance());
+        }
+        return INSTANCE;
+    }
+
+    /**
+     * Static helper to increments the counter for a specified event statistic on the singleton
+     *
+     * @param eventStatName The name of the event stat to increment
+     */
+    public static void increment(EventStatName eventStatName) {
+        instance().inc(eventStatName);
+    }
+
+    /**
+     * Constructs a new EventStatsManager with the provided settings accessor.
+     * Initializes statistics counters for all defined event types.
+     *
+     * @param settingsAccessor The accessor used to retrieve neural search settings
+     */
+    public EventStatsManager(NeuralSearchSettingsAccessor settingsAccessor) {
+        this.settingsAccessor = settingsAccessor;
+        this.stats = new HashMap<>();
+
+        for (EventStatName eventStatName : EnumSet.allOf(EventStatName.class)) {
+            if (eventStatName.getStatType() == EventStatType.TIMESTAMPED_COUNTER) {
+                stats.put(eventStatName, new TimestampedEventStat(eventStatName));
+            }
+        }
+    }
+
+    /**
+     *  Instance level method to increment the counter for a specified event statistic.
+     *
+     * @param eventStatName The name of the event stat to increment
+     */
+    public void inc(EventStatName eventStatName) {
+        boolean enabled = settingsAccessor.getIsStatsEnabled();
+        if (enabled && stats.containsKey(eventStatName)) {
```

Comment:
This could become an issue when there are many `eventStatName` values. While `HashMap.get` is generally constant time, its performance may degrade as the number of items grows. Given that this method is called frequently, we may need to explore an alternative approach to access the instance directly for better efficiency.

Reply:
Sure, I can have the EventStat be a member of the EventStatName enum and be instantiated in the enum constructor. That way we can access it directly without using a map at all.

If we have performance issues with hash map get, though, we do have other options, my understanding is the O(n) worst case happens when there are key object hash code collisions leading to putting objects in the same bucket and requiring a search on the bucket. There are ways we can mitigate this like implementing our hash code methods to ensure uniqueness, and Java has its own optimizations to address this like [storing collisions in a tree rather than linked list](https://openjdk.org/jeps/180).

I'll update again once I finish running performance benchmarks

Reply:
Yes. Storing it on a tree is also not O(1). Also, even if the hashcode are distinct, I don't think they will all have their own bucket. The number of bucket can be smaller then the number of hashcode.